source/opt/graphics_robust_access_pass.cpp

```diff
@@ -0,0 +1,984 @@
+                     "capability";
+  }
+
+  {
+    auto* inst = context()->module()->GetMemoryModel();
+    const auto addressing_model = inst->GetSingleWordOperand(0);
+    if (addressing_model != SpvAddressingModelLogical)
+      return Fail() << "Addressing model must be Logical.  Found "
+                    << inst->PrettyPrint();
+  }
+  return SPV_SUCCESS;
+}
+
+spv_result_t GraphicsRobustAccessPass::ProcessCurrentModule() {
+  auto err = IsCompatibleModule();
+  if (err != SPV_SUCCESS) return err;
+
+  ProcessFunction fn = [this](opt::Function* f) { return ProcessAFunction(f); };
+  module_status_.modified |= context()->ProcessReachableCallTree(fn);
+
+  // Need something here.  It's the price we pay for easier failure paths.
+  return SPV_SUCCESS;
+}
+
+bool GraphicsRobustAccessPass::ProcessAFunction(opt::Function* function) {
+  // Ensure that all pointers computed inside a function are within bounds.
+  // Find the access chains in this block before trying to modify them.
+  std::vector<Instruction*> access_chains;
+  std::vector<Instruction*> image_texel_pointers;
+  for (auto& block : *function) {
+    for (auto& inst : block) {
+      switch (inst.opcode()) {
+        case SpvOpAccessChain:
+        case SpvOpInBoundsAccessChain:
+          access_chains.push_back(&inst);
+          break;
+        case SpvOpImageTexelPointer:
+          image_texel_pointers.push_back(&inst);
+          break;
+        default:
+          break;
+      }
+    }
+  }
+  for (auto* inst : access_chains) {
+    ClampIndicesForAccessChain(inst);
+  }
+
+  for (auto* inst : image_texel_pointers) {
+    if (SPV_SUCCESS != ClampCoordinateForImageTexelPointer(inst)) break;
+  }
+  return module_status_.modified;
+}
+
+void GraphicsRobustAccessPass::ClampIndicesForAccessChain(
+    Instruction* access_chain) {
+  Instruction& inst = *access_chain;
+
+  auto* constant_mgr = context()->get_constant_mgr();
+  auto* def_use_mgr = context()->get_def_use_mgr();
+  auto* type_mgr = context()->get_type_mgr();
+
+  // Replaces one of the OpAccessChain index operands with a new value.
+  // Updates def-use analysis.
+  auto replace_index = [&inst, def_use_mgr](uint32_t operand_index,
+                                            Instruction* new_value) {
+    inst.SetOperand(operand_index, {new_value->result_id()});
+    def_use_mgr->AnalyzeInstUse(&inst);
+  };
+
+  // Replaces one of the OpAccesssChain index operands with a clamped value.
+  // Replace the operand at |operand_index| with
+  // the value computed from unsigned_max(%old_value_id, %max_value_id).
+  // It also Analyzes the definition and uses in the new instruction, and sets
+  // |merged| to true.
+  auto clamp_index = [&inst, this, &replace_index](
+                         uint32_t operand_index, Instruction* min_value,
+                         Instruction* old_value, Instruction* max_value) {
+    auto* clamp_inst = MakeClampInst(min_value, old_value, max_value, &inst);
```

Comment:
The GLSL.std450 instruction set defines UClamp operands as: `x`, `min`, `max`. Your operands look incorrectly ordered.